File: packages/sdk-ios/Sources/DecisioningSDK/DecisioningClient.swift
import CryptoKit
import Foundation

public enum InAppEventType: String, Codable, Sendable {
    case impression = "IMPRESSION"
    case click = "CLICK"
    case dismiss = "DISMISS"
}

public struct IdentityLookup: Codable, Sendable {
    public let attribute: String
    public let value: String

    public init(attribute: String, value: String) {
        self.attribute = attribute
        self.value = value
    }
}

public struct Tracking: Codable, Sendable {
    public let campaign_id: String
    public let message_id: String
    public let variant_id: String

    public init(campaign_id: String, message_id: String, variant_id: String) {
        self.campaign_id = campaign_id
        self.message_id = message_id
        self.variant_id = variant_id
    }
}

public enum JSONValue: Codable, Sendable, Equatable {
    case string(String)
    case number(Double)
    case bool(Bool)
    case object([String: JSONValue])
    case array([JSONValue])
    case null

    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if container.decodeNil() {
            self = .null
        } else if let value = try? container.decode(Bool.self) {
            self = .bool(value)
        } else if let value = try? container.decode(Double.self) {
            self = .number(value)
        } else if let value = try? container.decode(String.self) {
            self = .string(value)
        } else if let value = try? container.decode([String: JSONValue].self) {
            self = .object(value)
        } else if let value = try? container.decode([JSONValue].self) {
            self = .array(value)
        } else {
            throw DecodingError.dataCorruptedError(in: container, debugDescription: "Unsupported JSONValue")
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case .string(let value):
            try container.encode(value)
        case .number(let value):
            try container.encode(value)
        case .bool(let value):
            try container.encode(value)
        case .object(let value):
            try container.encode(value)
        case .array(let value):
            try container.encode(value)
        case .null:
            try container.encodeNil()
        }
    }
}

public struct DecideParams: Sendable {
    public let placement: String
    public let context: [String: JSONValue]?
    public let decisionKey: String?
    public let stackKey: String?

    public init(
        placement: String,
        context: [String: JSONValue]? = nil,
        decisionKey: String? = nil,
        stackKey: String? = nil
    ) {
        self.placement = placement
        self.context = context
        self.decisionKey = decisionKey
        self.stackKey = stackKey
    }
}

public struct DecideResponse: Codable, Sendable {
    public let show: Bool
    public let placement: String
    public let templateId: String
    public let ttl_seconds: Int
    public let tracking: Tracking
    public let payload: [String: JSONValue]
    public let debug: [String: JSONValue]?
}

public struct DecisioningConfig: Sendable {
    public let baseUrl: URL
    public let appKey: String
    public let bearerToken: String?
    public let apiKey: String?
    public let environment: String?
    public let defaultContext: [String: JSONValue]
    public let contextAllowlist: Set<String>
    public let decideTimeoutMs: TimeInterval
    public let eventsTimeoutMs: TimeInterval
    public let cacheTtlSeconds: Int
    public let staleTtlSeconds: Int
    public let decideRetryCount: Int
    public let debug: Bool

    public init(
        baseUrl: URL,
        appKey: String,
        bearerToken: String? = nil,
        apiKey: String? = nil,
        environment: String? = nil,
        defaultContext: [String: JSONValue] = [:],
        contextAllowlist: Set<String> = ["locale", "deviceType", "appVersion"],
        decideTimeoutMs: TimeInterval = 500,
        eventsTimeoutMs: TimeInterval = 1_000,
        cacheTtlSeconds: Int = 60,
        staleTtlSeconds: Int = 1_800,
        decideRetryCount: Int = 1,
        debug: Bool = false
    ) {
        self.baseUrl = baseUrl
        self.appKey = appKey
        self.bearerToken = bearerToken
        self.apiKey = apiKey
        self.environment = environment
        self.defaultContext = defaultContext
        self.contextAllowlist = contextAllowlist
        self.decideTimeoutMs = decideTimeoutMs
        self.eventsTimeoutMs = eventsTimeoutMs
        self.cacheTtlSeconds = cacheTtlSeconds
        self.staleTtlSeconds = staleTtlSeconds
        self.decideRetryCount = decideRetryCount
        self.debug = debug
    }
}

public struct EventTarget: Sendable {
    public let placement: String
    public let tracking: Tracking

    public init(placement: String, tracking: Tracking) {
        self.placement = placement
        self.tracking = tracking
    }
}

public struct CacheEntry: Codable, Sendable {
    public let response: DecideResponse
    public let expiresAt: Date
    public let staleUntil: Date
}

public protocol DecideCacheStore: Sendable {
    func get(key: String) async -> CacheEntry?
    func put(key: String, value: CacheEntry) async
}

public actor MemoryDecideCacheStore: DecideCacheStore {
    private var storage: [String: CacheEntry] = [:]

    public init() {}

    public func get(key: String) async -> CacheEntry? {
        storage[key]
    }

    public func put(key: String, value: CacheEntry) async {
        storage[key] = value
    }
}

public actor UserDefaultsDecideCacheStore: DecideCacheStore {
    private let defaults: UserDefaults
    private let prefix: String
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    public init(defaults: UserDefaults = .standard, prefix: String = "decisioning.decide") {
        self.defaults = defaults
        self.prefix = prefix
    }

    public func get(key: String) async -> CacheEntry? {
        guard let data = defaults.data(forKey: "\(prefix).\(key)") else {
            return nil
        }
        return try? decoder.decode(CacheEntry.self, from: data)
    }

    public func put(key: String, value: CacheEntry) async {
        guard let data = try? encoder.encode(value) else {
            return
        }
        defaults.set(data, forKey: "\(prefix).\(key)")
    }
}

public protocol EventQueueStore: Sendable {
    func enqueue(_ event: EventRequest) async
    func list() async -> [EventRequest]
    func remove(ids: Set<String>) async
}

public actor MemoryEventQueueStore: EventQueueStore {
    private var queue: [EventRequest] = []

    public init() {}

    public func enqueue(_ event: EventRequest) async {
        queue.append(event)
    }

    public func list() async -> [EventRequest] {
        queue
    }

    public func remove(ids: Set<String>) async {
        queue.removeAll { ids.contains($0.eventId) }
    }
}

public actor UserDefaultsEventQueueStore: EventQueueStore {
    private let defaults: UserDefaults
    private let key: String
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    public init(defaults: UserDefaults = .standard, key: String = "decisioning.eventQueue") {
        self.defaults = defaults
        self.key = key
    }

    public func enqueue(_ event: EventRequest) async {
        var queue = await list()
        queue.append(event)
        persist(queue)
    }

    public func list() async -> [EventRequest] {
        guard let data = defaults.data(forKey: key) else {
            return []
        }
        return (try? decoder.decode([EventRequest].self, from: data)) ?? []
    }

    public func remove(ids: Set<String>) async {
        let filtered = await list().filter { !ids.contains($0.eventId) }
        persist(filtered)
    }

    private func persist(_ queue: [EventRequest]) {
        guard let data = try? encoder.encode(queue) else {
            return
        }
        defaults.set(data, forKey: key)
    }
}

public struct EventRequest: Codable, Sendable {
    public let eventType: InAppEventType
    public let ts: String
    public let appKey: String
    public let placement: String
    public let tracking: Tracking
    public let profileId: String?
    public let lookup: IdentityLookup?
    public let context: [String: JSONValue]?
    public let eventId: String
}

public actor DecisioningClient {
    private let config: DecisioningConfig
    private let session: URLSession
    private let cacheStore: DecideCacheStore
    private let queueStore: EventQueueStore
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    private var profileId: String?
    private var anonymousId: String?
    private var lookup: IdentityLookup?

    public init(
        config: DecisioningConfig,
        session: URLSession = .shared,
        cacheStore: DecideCacheStore = MemoryDecideCacheStore(),
        queueStore: EventQueueStore = MemoryEventQueueStore()
    ) {
        self.config = config
        self.session = session
        self.cacheStore = cacheStore
        self.queueStore = queueStore
    }

    public func setProfileId(_ value: String) {
        profileId = value
    }

    public func setAnonymousId(_ value: String) {
        anonymousId = value
    }

    public func setLookup(attribute: String, value: String) {
        lookup = IdentityLookup(attribute: attribute, value: value)
    }

    public func decide(_ params: DecideParams) async throws -> DecideResponse {
        let mergedContext = config.defaultContext.merging(params.context ?? [:], uniquingKeysWith: { _, rhs in rhs })
        let key = buildCacheKey(placement: params.placement, context: mergedContext)

        if let cached = await cacheStore.get(key: key), cached.expiresAt > Date() {
            return cached.response
        }

        do {
            let response = try await decideWithRetry(params: params, mergedContext: mergedContext, retries: config.decideRetryCount)
            let ttl = max(1, response.ttl_seconds > 0 ? response.ttl_seconds : config.cacheTtlSeconds)
            let now = Date()
            let entry = CacheEntry(
                response: response,
                expiresAt: now.addingTimeInterval(Double(ttl)),
                staleUntil: now.addingTimeInterval(Double(ttl + config.staleTtlSeconds))
            )
            await cacheStore.put(key: key, value: entry)
            return response
        } catch {
            if let stale = await cacheStore.get(key: key), stale.staleUntil > Date() {
                return stale.response
            }
            throw error
        }
    }

    public func trackImpression(_ target: DecideResponse, context: [String: JSONValue] = [:]) async throws {
        try await track(.impression, target: EventTarget(placement: target.placement, tracking: target.tracking), context: context)
    }

    public func trackClick(_ target: DecideResponse, context: [String: JSONValue] = [:]) async throws {
        try await track(.click, target: EventTarget(placement: target.placement, tracking: target.tracking), context: context)
    }

    public func trackDismiss(_ target: DecideResponse, context: [String: JSONValue] = [:]) async throws {
        try await track(.dismiss, target: EventTarget(placement: target.placement, tracking: target.tracking), context: context)
    }

    public func flushEvents() async throws {
        let snapshot = await queueStore.list()
        var sent = Set<String>()

        for event in snapshot {
            if try await sendEventWithRetry(event) {
                sent.insert(event.eventId)
            }
        }

        if !sent.isEmpty {
            await queueStore.remove(ids: sent)
        }
    }

    public func pendingEventCount() async -> Int {
        await queueStore.list().count
    }

    internal func buildCacheKey(placement: String, context: [String: JSONValue]) -> String {
        let identity: String
        if let profileId {
            identity = "profile:\(profileId)"
        } else if let lookup {
            identity = "lookup:\(lookup.attribute):\(lookup.value)"
        } else if let anonymousId {
            identity = "anonymous:\(anonymousId)"
        } else {
            identity = "identity:unknown"
        }

        let allowed = context
            .filter { config.contextAllowlist.contains($0.key) }
            .sorted { $0.key < $1.key }
            .map { "\($0.key)=\($0.value)" }
            .joined(separator: "&")

        return sha256Hex("\(config.appKey):\(placement):\(identity):\(allowed)")
    }

    private func decideWithRetry(params: DecideParams, mergedContext: [String: JSONValue], retries: Int) async throws -> DecideResponse {
        do {
            return try await requestDecide(params: params, mergedContext: mergedContext)
        } catch {
            if retries <= 0 {
                throw error
            }
            return try await requestDecide(params: params, mergedContext: mergedContext)
        }
    }

    private func requestDecide(params: DecideParams, mergedContext: [String: JSONValue]) async throws -> DecideResponse {
        let payload = DecideRequestPayload(
            appKey: config.appKey,
            placement: params.placement,
            decisionKey: params.decisionKey,
            stackKey: params.stackKey,
            profileId: profileId ?? anonymousId,
            lookup: lookup,
            context: mergedContext
        )

        let request = try buildRequest(path: "/v2/inapp/decide", body: payload, timeoutMs: config.decideTimeoutMs)
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse, 200..<300 ~= http.statusCode else {
            throw NSError(domain: "DecisioningClient", code: 1)
        }

        return try decoder.decode(DecideResponse.self, from: data)
    }

    private func track(_ type: InAppEventType, target: EventTarget, context: [String: JSONValue]) async throws {
        let timestamp = ISO8601DateFormatter().string(from: Date())
        let bucket = Int(Date().timeIntervalSince1970 / 60)
        let eventId = sha256Hex("\(target.tracking.message_id):\(type.rawValue):\(bucket)")

        let payload = EventRequest(
            eventType: type,
            ts: timestamp,
            appKey: config.appKey,
            placement: target.placement,
            tracking: target.tracking,
            profileId: profileId ?? anonymousId,
            lookup: lookup,
            context: config.defaultContext.merging(context, uniquingKeysWith: { _, rhs in rhs }),
            eventId: eventId
        )

        await queueStore.enqueue(payload)
        try await flushEvents()
    }

    private func sendEventWithRetry(_ payload: EventRequest) async throws -> Bool {
        let request = try buildRequest(path: "/v2/inapp/events", body: payload, timeoutMs: config.eventsTimeoutMs, eventId: payload.eventId)

        let first = try await session.data(for: request)
        if let response = first.1 as? HTTPURLResponse {
            if 200..<300 ~= response.statusCode {
                return true
            }
            if response.statusCode < 500 {
                return false
            }
        }

        try await Task.sleep(nanoseconds: 150_000_000)
        let second = try await session.data(for: request)
        if let response = second.1 as? HTTPURLResponse {
            return 200..<300 ~= response.statusCode
        }
        return false
    }

    internal func buildRequest<T: Encodable>(
        path: String,
        body: T,
        timeoutMs: TimeInterval,
        eventId: String? = nil
    ) throws -> URLRequest {
        let url = config.baseUrl.appendingPathComponent(path.replacingOccurrences(of: "^/", with: "", options: .regularExpression))
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = timeoutMs / 1000
        request.httpBody = try encoder.encode(body)
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(UUID().uuidString, forHTTPHeaderField: "X-Request-Id")

        if let environment = config.environment {
            request.setValue(environment, forHTTPHeaderField: "X-ENV")
        }
        if let bearerToken = config.bearerToken {
            request.setValue("Bearer \(bearerToken)", forHTTPHeaderField: "Authorization")
        } else if let apiKey = config.apiKey {
            request.setValue(apiKey, forHTTPHeaderField: "X-API-KEY")
        }
        if let eventId {
            request.setValue(eventId, forHTTPHeaderField: "X-Event-Id")
        }
        return request
    }
}

private struct DecideRequestPayload: Encodable {
    let appKey: String
    let placement: String
    let decisionKey: String?
    let stackKey: String?
    let profileId: String?
    let lookup: IdentityLookup?
    let context: [String: JSONValue]?
}

private func sha256Hex(_ string: String) -> String {
    let digest = SHA256.hash(data: Data(string.utf8))
    return digest.map { String(format: "%02x", $0) }.joined()
}

File: packages/sdk-android/src/main/kotlin/com/decisioning/sdk/DecisioningClient.kt
package com.decisioning.sdk

import java.io.File
import java.security.MessageDigest
import java.time.Instant
import java.util.Base64
import java.util.UUID
import kotlin.math.max
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

@Serializable
data class IdentityLookup(
    val attribute: String,
    val value: String
)

@Serializable
data class DecideRequest(
    val placement: String,
    val context: Map<String, JsonElement>? = null,
    val decisionKey: String? = null,
    val stackKey: String? = null
)

@Serializable
data class Tracking(
    @SerialName("campaign_id") val campaignId: String,
    @SerialName("message_id") val messageId: String,
    @SerialName("variant_id") val variantId: String,
    @SerialName("experiment_id") val experimentId: String? = null,
    @SerialName("experiment_version") val experimentVersion: Int? = null,
    @SerialName("is_holdout") val isHoldout: Boolean? = null,
    @SerialName("allocation_id") val allocationId: String? = null
)

@Serializable
data class DecideResponse(
    val show: Boolean,
    val placement: String,
    val templateId: String = "none",
    @SerialName("ttl_seconds") val ttlSeconds: Long = 0,
    val tracking: Tracking,
    val payload: Map<String, JsonElement> = emptyMap(),
    val debug: Map<String, JsonElement> = emptyMap()
)

enum class InAppEventType {
    IMPRESSION,
    CLICK,
    DISMISS
}

@Serializable
data class EventRequest(
    val eventType: String,
    val ts: String,
    val appKey: String,
    val placement: String,
    val tracking: Tracking,
    val profileId: String? = null,
    val lookup: IdentityLookup? = null,
    val context: Map<String, JsonElement>? = null,
    val eventId: String
)

@Serializable
private data class DecideRequestBodyPayload(
    val appKey: String,
    val placement: String,
    val decisionKey: String? = null,
    val stackKey: String? = null,
    val profileId: String? = null,
    val anonymousId: String? = null,
    val lookup: IdentityLookup? = null,
    val context: Map<String, JsonElement>? = null
)

data class DecisioningConfig(
    val baseUrl: String,
    val appKey: String,
    val environment: String? = null,
    val bearerToken: String? = null,
    val apiKey: String? = null,
    val defaultContext: Map<String, JsonElement> = emptyMap(),
    val contextAllowlist: Set<String> = setOf("locale", "deviceType", "appVersion"),
    val decideTimeoutMs: Long = 500,
    val eventsTimeoutMs: Long = 1000,
    val cacheTtlSeconds: Long = 60,
    val staleTtlSeconds: Long = 1800,
    val debug: Boolean = false,
    val decideRetryCount: Int = 1
)

@Serializable
data class CacheEntry(
    val response: DecideResponse,
    val expiresAtMs: Long,
    val staleUntilMs: Long
)

interface DecideCacheStore {
    suspend fun get(key: String): CacheEntry?
    suspend fun put(key: String, entry: CacheEntry)
}

class InMemoryDecideCacheStore : DecideCacheStore {
    private val entries = LinkedHashMap<String, CacheEntry>()
    private val mutex = Mutex()

    override suspend fun get(key: String): CacheEntry? = mutex.withLock { entries[key] }

    override suspend fun put(key: String, entry: CacheEntry) {
        mutex.withLock {
            entries[key] = entry
            if (entries.size > 128) {
                val first = entries.keys.firstOrNull()
                if (first != null) {
                    entries.remove(first)
                }
            }
        }
    }
}

interface KeyValueStore {
    suspend fun getString(key: String): String?
    suspend fun putString(key: String, value: String)
    suspend fun remove(key: String)
}

class InMemoryKeyValueStore : KeyValueStore {
    private val map = mutableMapOf<String, String>()
    private val mutex = Mutex()

    override suspend fun getString(key: String): String? = mutex.withLock { map[key] }

    override suspend fun putString(key: String, value: String) {
        mutex.withLock { map[key] = value }
    }

    override suspend fun remove(key: String) {
        mutex.withLock { map.remove(key) }
    }
}

class FileKeyValueStore(private val file: File) : KeyValueStore {
    private val mutex = Mutex()

    override suspend fun getString(key: String): String? = mutex.withLock {
        readState()[key]
    }

    override suspend fun putString(key: String, value: String) {
        mutex.withLock {
            val state = readState().toMutableMap()
            state[key] = value
            writeState(state)
        }
    }

    override suspend fun remove(key: String) {
        mutex.withLock {
            val state = readState().toMutableMap()
            state.remove(key)
            writeState(state)
        }
    }

    private fun readState(): Map<String, String> {
        if (!file.exists()) {
            return emptyMap()
        }
        val raw = file.readText().trim()
        if (raw.isEmpty()) {
            return emptyMap()
        }
        return raw.lineSequence()
            .mapNotNull { line ->
                val idx = line.indexOf("=")
                if (idx <= 0) {
                    null
                } else {
                    val decoded = runCatching {
                        String(Base64.getDecoder().decode(line.substring(idx + 1)))
                    }.getOrNull()
                    if (decoded == null) {
                        null
                    } else {
                        line.substring(0, idx) to decoded
                    }
                }
            }
            .toMap()
    }

    private fun writeState(state: Map<String, String>) {
        file.parentFile?.mkdirs()
        val content = state.entries.joinToString("\n") {
            "${it.key}=${Base64.getEncoder().encodeToString(it.value.toByteArray())}"
        }
        file.writeText(content)
    }
}

open class PreferencesDecideCacheStore(
    private val store: KeyValueStore,
    private val namespace: String = "decisioning:decide",
    private val json: Json = Json { ignoreUnknownKeys = true }
) : DecideCacheStore {
    override suspend fun get(key: String): CacheEntry? {
        val raw = store.getString("$namespace:$key") ?: return null
        return runCatching { json.decodeFromString<CacheEntry>(raw) }.getOrNull()
    }

    override suspend fun put(key: String, entry: CacheEntry) {
        store.putString("$namespace:$key", json.encodeToString(entry))
    }
}

class FileDecideCacheStore(
    file: File,
    namespace: String = "decisioning:decide",
    json: Json = Json { ignoreUnknownKeys = true }
) : PreferencesDecideCacheStore(
    store = FileKeyValueStore(file),
    namespace = namespace,
    json = json
)

interface EventQueueStore {
    suspend fun add(event: EventRequest)
    suspend fun list(): List<EventRequest>
    suspend fun remove(eventIds: Set<String>)
}

class InMemoryEventQueueStore : EventQueueStore {
    private val queue = mutableListOf<EventRequest>()
    private val mutex = Mutex()

    override suspend fun add(event: EventRequest) {
        mutex.withLock { queue.add(event) }
    }

    override suspend fun list(): List<EventRequest> = mutex.withLock { queue.toList() }

    override suspend fun remove(eventIds: Set<String>) {
        mutex.withLock {
            queue.removeAll { it.eventId in eventIds }
        }
    }
}

open class PreferencesEventQueueStore(
    private val store: KeyValueStore,
    private val key: String = "decisioning:event-queue",
    private val json: Json = Json { ignoreUnknownKeys = true; encodeDefaults = true }
) : EventQueueStore {
    private val mutex = Mutex()

    override suspend fun add(event: EventRequest) {
        mutex.withLock {
            val queue = loadQueue().toMutableList()
            queue.add(event)
            store.putString(key, json.encodeToString(queue))
        }
    }

    override suspend fun list(): List<EventRequest> = mutex.withLock { loadQueue() }

    override suspend fun remove(eventIds: Set<String>) {
        mutex.withLock {
            val queue = loadQueue().filterNot { it.eventId in eventIds }
            store.putString(key, json.encodeToString(queue))
        }
    }

    private suspend fun loadQueue(): List<EventRequest> {
        val raw = store.getString(key) ?: return emptyList()
        return runCatching { json.decodeFromString<List<EventRequest>>(raw) }.getOrElse { emptyList() }
    }
}

class FileEventQueueStore(
    file: File,
    key: String = "decisioning:event-queue",
    json: Json = Json { ignoreUnknownKeys = true; encodeDefaults = true }
) : PreferencesEventQueueStore(
    store = FileKeyValueStore(file),
    key = key,
    json = json
)

class DecisioningClient(
    private val config: DecisioningConfig,
    private val cacheStore: DecideCacheStore = InMemoryDecideCacheStore(),
    private val queueStore: EventQueueStore = InMemoryEventQueueStore(),
    private val json: Json = Json { ignoreUnknownKeys = true; encodeDefaults = true },
    private val client: OkHttpClient = OkHttpClient(),
    private val nowMs: () -> Long = { System.currentTimeMillis() }
) {
    private var profileId: String? = null
    private var anonymousId: String? = null
    private var lookup: IdentityLookup? = null

    fun setProfileId(value: String) {
        profileId = value
    }

    fun setAnonymousId(value: String) {
        anonymousId = value
    }

    fun setLookup(attribute: String, value: String) {
        lookup = IdentityLookup(attribute = attribute, value = value)
    }

    suspend fun decide(input: DecideRequest): DecideResponse {
        val mergedContext = config.defaultContext + (input.context ?: emptyMap())
        val cacheKey = buildCacheKey(input.placement, mergedContext)
        val now = nowMs()

        val fresh = cacheStore.get(cacheKey)
        if (fresh != null && fresh.expiresAtMs > now) {
            return fresh.response
        }

        return try {
            val response = decideWithRetry(input, mergedContext, config.decideRetryCount)
            val ttl = if (response.ttlSeconds > 0) response.ttlSeconds else config.cacheTtlSeconds
            val entry = CacheEntry(
                response = response,
                expiresAtMs = now + ttl * 1_000,
                staleUntilMs = now + (ttl + config.staleTtlSeconds) * 1_000
            )
            cacheStore.put(cacheKey, entry)
            response
        } catch (error: Throwable) {
            val stale = cacheStore.get(cacheKey)
            if (stale != null && stale.staleUntilMs > nowMs()) {
                stale.response
            } else {
                throw error
            }
        }
    }

    suspend fun trackImpression(target: DecideResponse, context: Map<String, JsonElement> = emptyMap()) {
        track(InAppEventType.IMPRESSION, target, context)
    }

    suspend fun trackClick(target: DecideResponse, context: Map<String, JsonElement> = emptyMap()) {
        track(InAppEventType.CLICK, target, context)
    }

    suspend fun trackDismiss(target: DecideResponse, context: Map<String, JsonElement> = emptyMap()) {
        track(InAppEventType.DISMISS, target, context)
    }

    suspend fun flushEvents() {
        val snapshot = queueStore.list()
        if (snapshot.isEmpty()) {
            return
        }

        val sent = mutableSetOf<String>()
        for (event in snapshot) {
            val success = sendEventWithRetry(event)
            if (success) {
                sent.add(event.eventId)
            }
        }

        if (sent.isNotEmpty()) {
            queueStore.remove(sent)
        }
    }

    suspend fun pendingEventCount(): Int = queueStore.list().size

    internal fun buildCacheKey(placement: String, context: Map<String, JsonElement>): String {
        val identity = when {
            !profileId.isNullOrBlank() -> "profile:$profileId"
            lookup != null -> "lookup:${lookup?.attribute}:${lookup?.value}"
            !anonymousId.isNullOrBlank() -> "anonymous:$anonymousId"
            else -> "identity:unknown"
        }
        val allowed = context.filterKeys { it in config.contextAllowlist }.toSortedMap()
        return "${config.appKey}:$placement:$identity:$allowed".sha256Hex()
    }

    private suspend fun decideWithRetry(
        input: DecideRequest,
        mergedContext: Map<String, JsonElement>,
        retries: Int
    ): DecideResponse {
        return try {
            requestDecide(input, mergedContext)
        } catch (error: Throwable) {
            if (retries <= 0 || !isNetworkError(error)) {
                throw error
            }
            requestDecide(input, mergedContext)
        }
    }

    private suspend fun requestDecide(input: DecideRequest, mergedContext: Map<String, JsonElement>): DecideResponse {
        val body = DecideRequestBodyPayload(
            appKey = config.appKey,
            placement = input.placement,
            decisionKey = input.decisionKey,
            stackKey = input.stackKey,
            profileId = profileId,
            anonymousId = anonymousId,
            lookup = lookup,
            context = mergedContext
        )

        val request = buildRequest("/v2/inapp/decide", json.encodeToString(body), config.decideTimeoutMs)
        val response = execute(request)
        if (!response.isSuccessful) {
            throw RuntimeException("decide failed: ${response.code}")
        }

        val payload = response.body?.string().orEmpty()
        return json.decodeFromString(payload)
    }

    private suspend fun track(type: InAppEventType, target: DecideResponse, context: Map<String, JsonElement>) {
        val ts = Instant.ofEpochMilli(nowMs()).toString()
        val tsBucket = max(1, nowMs() / 60_000)
        val eventId = "${target.tracking.messageId}:${type.name}:$tsBucket".sha256Hex()
        val event = EventRequest(
            eventType = type.name,
            ts = ts,
            appKey = config.appKey,
            placement = target.placement,
            tracking = target.tracking,
            profileId = profileId ?: anonymousId,
            lookup = lookup,
            context = config.defaultContext + context,
            eventId = eventId
        )

        queueStore.add(event)
        flushEvents()
    }

    private suspend fun sendEventWithRetry(event: EventRequest): Boolean {
        val request = buildRequest("/v2/inapp/events", json.encodeToString(event), config.eventsTimeoutMs, event.eventId)
        val first = execute(request)
        if (first.isSuccessful) {
            return true
        }

        if (first.code < 500) {
            return false
        }

        delay(150)
        val second = execute(request)
        return second.isSuccessful
    }

    private fun buildRequest(path: String, payloadJson: String, timeoutMs: Long, eventId: String? = null): Request {
        val builder = Request.Builder()
            .url("${config.baseUrl.removeSuffix("/")}$path")
            .post(payloadJson.toRequestBody("application/json".toMediaType()))
            .header("Content-Type", "application/json")
            .header("X-Request-Id", UUID.randomUUID().toString())

        if (!config.environment.isNullOrBlank()) {
            builder.header("X-ENV", config.environment)
        }
        if (!config.bearerToken.isNullOrBlank()) {
            builder.header("Authorization", "Bearer ${config.bearerToken}")
        } else if (!config.apiKey.isNullOrBlank()) {
            builder.header("X-API-KEY", config.apiKey)
        }
        if (!eventId.isNullOrBlank()) {
            builder.header("X-Event-Id", eventId)
        }

        return builder.tag(Long::class.java, timeoutMs).build()
    }

    private suspend fun execute(request: Request) = withContext(Dispatchers.IO) {
        val timeoutMs = request.tag(Long::class.java) ?: config.eventsTimeoutMs
        val scoped = client.newBuilder().callTimeout(java.time.Duration.ofMillis(timeoutMs)).build()
        scoped.newCall(request).execute()
    }

    private fun isNetworkError(error: Throwable): Boolean {
        return error is java.io.IOException
    }
}

private fun String.sha256Hex(): String {
    val bytes = MessageDigest.getInstance("SHA-256").digest(toByteArray())
    return bytes.joinToString("") { "%02x".format(it) }
}
